src/engine.ml

```diff
@@ -1332,9 +1332,40 @@ let wrap_hmac_control now ts mtu session tls_auth key transport outs =
   in
   (session, transport, List.rev outs)
 
-let wrap_tls_crypt_control now ts mtu session tls_crypt key transport outs =
+let wrap_tls_crypt_control now ts mtu session (tls_crypt, wkc) needs_wkc key
+    transport outs =
   let now_ts = ptime_to_ts_exn now in
-  let outs = split_control mtu outs in
+  (* If we reply with hmac cookie we must split such that the first control
+     packet, the Control_wkc has room for the /cleartext/ wkc, and fix the
+     packet length afterwards *)
+  let session, transport, maybe_out, outs =
+    match (needs_wkc, outs) with
+    | true, (`Control, data) :: rest ->
+        let l = min (mtu - Cstruct.length wkc) (Cstruct.length data) in
+        let data, data' = Cstruct.split data l in
+        let rest =
+          if Cstruct.is_empty data' then rest else (`Control, data') :: rest
+        in
+        let session, transport, header = header session transport now_ts in
+        let transport, m_id = next_message_id transport in
+        let p = `Control (Packet.Control_wkc, (header, m_id, data)) in
+        (* First we encrypt *)
+        let out = encrypt_and_out session.protocol tls_crypt (key, p) in
+        (* Then we append wkc and fix the length if TCP *)
+        let len =
+          Cstruct.length out
+          - match session.protocol with `Tcp -> 2 | `Udp -> 0
+        in
+        let out = Cstruct.append out wkc in
+        let proto = Packet.encode_protocol session.protocol len in
+        Cstruct.blit proto 0 out 0 (Cstruct.length proto);
```

Comment:
I'm not too fond of this hack. We can't do the same hack as for `HARD_RESET_CLIENT_V3` where we pretend the wKc is part of the packet data and special case the encoder. On the other hand maybe we should implement the same hack for `HARD_RESET_CLIENT_V3` instead of using two different hacks.

Reply:
After some thinking I think it would be interesting to always send a `CONTROL_WKC_V1` with an empty body and then pretend the `wKc` is the body and employ the same hack we use for `HARD_RESET_CLIENT_V3`.